alts/src/main/java/io/grpc/alts/GoogleDefaultChannelCredentials.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.alts;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.collect.ImmutableList;
+import io.grpc.CallCredentials;
+import io.grpc.ChannelCredentials;
+import io.grpc.CompositeChannelCredentials;
+import io.grpc.ExperimentalApi;
+import io.grpc.Status;
+import io.grpc.alts.internal.AltsProtocolNegotiator.GoogleDefaultProtocolNegotiatorFactory;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.internal.SharedResourcePool;
+import io.grpc.netty.GrpcSslContexts;
+import io.grpc.netty.InternalNettyChannelCredentials;
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.netty.handler.ssl.SslContext;
+import java.io.IOException;
+import javax.net.ssl.SSLException;
+
+/**
+ * Google default credentials. This class sets up a secure channel using ALTS if applicable and
+ * using TLS as fallback.
+ */
+@ExperimentalApi("TODO")
+public final class GoogleDefaultChannelCredentials {
+  private GoogleDefaultChannelCredentials() {}
+
+  /**
+   * Creates Google default credentials uses a secure channel with ALTS if applicable and uses TLS
+   * as fallback.
+   */
+  public static ChannelCredentials create() {
+    ChannelCredentials googleDefaultNegotiator =
+        InternalNettyChannelCredentials.create(createClientFactory());
+    CallCredentials callCredentials;
+    try {
+      callCredentials = MoreCallCredentials.from(GoogleCredentials.getApplicationDefault());
+    } catch (IOException e) {
+      // TODO(ejona): Should this just throw?
```

Comment:
As I commented elsewhere throwing sounds better than `FailingCallCredentials` unless you want to match existing behavior.

Reply:
The other case is a bit different. We can check if we're on GCP by looking at the BIOS information, which would be pretty quick and reliable. And ComputeEngineCredentials doesn't do any I/O when created.

In this case `getApplicationDefault()` itself does network I/O and can fail for reasons other than misconfiguration. It seems wrong to permanently break a channel in the case of that failure without providing a good way for users to manage the error themselves. So that's why I suggested throwing here.

Actually, I guess there is another option here: create CallCredentials that call `getApplicationDefault()` as part of its loading of credentials and if it fails it could continue retrying.

In either case, though, I don't think it is appropriate to change the behavior in this PR. And we probably can't change the existing builder. But we do really want to decide the API for GoogleDefaultChannelCredentials before stabilizing it.

Reply:
It looks like we shall also call `CheckGcpEnvironment.isOnGcp()` for GoogleDefaultChannelCredentials.

Do we expect caller of GoogleDefaultChannelCredentials outside GCP? It is not safe to call ALTS without GCP platform check. 

If we do plan to support GoogleDefaultChannelCredentials outside GCP, we can do platform check, if running outside GCP, disable ALTS, only TLS is allowed. WDYT?

Reply:
> Do we expect caller of GoogleDefaultChannelCredentials outside GCP? It is not safe to call ALTS without GCP platform check.

Absolutely we do. In that case we won't get any gRPCLB addresses and so won't trigger the ALTS code path.

The entire point of GoogleDefaultChannelCredentials (which mirrors C core) is to auto-select the proper credentials to use when communicating with Google independent of where you are. Just like `GoogleCredentials.getApplicationDefault()`.

In any case, I do not want to change existing behavior in this PR.

Reply:
IIUC, google_default_credentials can only be run on GCP. Agree this PR should not change existing behavior. 
It is reasonable to not perform GCP check for GoogleDefaultCredentials, as we trust DNS to signal if we enable ALTS or not. If GDC runs outside GCP, ALTS code path won't be triggered.

Reply:
`ComputeEngineChannelCredentials` is the one restricted to GCP. No reason to restrict `GoogleDefaultChannelCredentials` to GCP (similar to the behavior of `GoogleDefaultChannelBuilder`)